src/middle/Stmt.ml

```diff
@@ -115,26 +115,40 @@ module Numbered = struct
 end
 
 module Helpers = struct
+  let temp_vars exprs : Located.t list * Expr.Typed.t list * (unit -> unit) =
+    let sym, reset = Gensym.enter () in
+    let rec loop es sym inits vars =
+      match es with
+      | [] -> (inits, vars)
+      | e :: es -> (
+        match e with
+        | Expr.{Fixed.pattern= Var _; _} -> loop es sym inits (e :: vars)
+        | _ ->
```

Comment:
can you combine these `match`es? Like
```ocaml
match es with
| [] -> (inits, vars)
| Expr{Fixed.pattern=Var _;_} as e :: es -> loop es sym inits (e :: vars)
| e :: es -> ...
```
Also, I'm not sure if that should be `pattern=Var _` or `pattern=Var _ | Indexed _`? Does `Indexed` as an rvalue create copies or references?

Reply:
I believe this would create copies. The rvalue in C++ ends up being something like `Eigen::Block`, but when that is assigned to an `Eigen::Matrix` it would get evaluated.

That said, and `Indexed` could very well be something crazy like `(a +  b * X)[:]`, so I think we want anything other than a Var here to have a temporary created.